Add Cloud SQL guide (#6349)
gravitational/teleport: docs/pages/database-access/guides/postgres-cloudsql.mdx

![Check IAM Authentication](../../../img/database-access/guides/cloudsql/iam-existing-instance.png)

If it isn't enabled, you can add this flag using the "Edit configuration" dialog
at the bottom of the Configuration panel. Note, changing this setting may
require a database instance reboot.

## Create Service Account for Database

Teleport uses service accounts to connect to Cloud SQL databases.

Go to the IAM & Admin [Service Accounts](https://console.cloud.google.com/iam-admin/serviceaccounts)
page and create a new service account:

![Create Service Account](../../../img/database-access/guides/cloudsql/service-account-db.png)

Press "Create".

On the second step grant this service account the "Cloud SQL Instance User" role
which will allow it to connect to Cloud SQL instances using IAM token for
authentication:

![Grant Cloud SQL Instance User to Service Account](../../../img/database-access/guides/cloudsql/service-account-db-grant.png)

Press "Done".

Now go back to the Users page of your Cloud SQL instance and add a new user
account. In the sidebar, choose "Cloud IAM" authentication type and add the
service account you've just created:

> ![Add Cloud SQL User Account](../../../img/database-access/guides/cloudsql/add-user-account.png)

Reviewer: Any reason for adding `teleport@` vs just `teleport`?

Author: We need to add service account as a database user, a service account name is an email.

Overview page:

![Instance ID](../../../img/database-access/guides/cloudsql/instance-id.png)

- Cloud SQL instance endpoint.

You will use the instance's public IP address to connect to it. It can be viewed
on the "Connect to this instance" panel on the Overview page:

![Instance Public IP](../../../img/database-access/guides/cloudsql/instance-public-ip.png)

- Cloud SQL instance root certificate.

The instance's root certificate is required so Teleport can validate the
certificate presented by the database instance. You can download `server-ca.pem`
file from the Connections tab under Security section:

![Instance Root Certificate](../../../img/database-access/guides/cloudsql/instance-root-ca.png)

## Configure Teleport

Cloud SQL support for Teleport Database Access is available starting from
`6.1` release.

Download the appropriate version of Teleport for your platform from
our [downloads page](https://goteleport.com/teleport/download).

Follow the installation [instructions](../../installation.mdx).

### Start Auth/Proxy Service

Create a configuration file for a Teleport service that will be running
auth and proxy servers:

> ```yaml

Reviewer 2: Please use `teleport configure` command and ACME for this getting started guide.

Author: Done.

teleport:
  data_dir: /var/lib/teleport
  nodename: test
auth_service:
  enabled: "yes"
proxy_service:
  enabled: "yes"
  # Set public address proxy will be reachable at.
  public_addr: teleport.example.com:3080
  # The PostgreSQL proxy listens on the regular web proxy port.
  web_listen_addr: 0.0.0.0:3080
ssh_service:
  enabled: "no"

> ```

Reviewer: I would suggest we use 'Getting started with Teleport' as the canonical reference for setting up Teleport.

Author: These are supposed to be self-contained guides, incl. setting up Teleport. This would also make it inconsistent with the other guides.

Reviewer: In that case, I would suggest we need to setup TLS for the proxy.

Author: Yeah, we can use teleport configure and ACME here like in quickstart. I'll update.

Author: Updated.

Start the service:

```sh
$ teleport start --config=/path/to/teleport.yaml
```

Generate a short-lived join token for the database service and save it for
example in `/tmp/token`:

```sh
$ tctl tokens add \
    --type=db \
    --db-name=test \
    --db-protocol=postgres \
    --db-uri=35.1.2.3:5432
```

### Create Role and User

> Create the role that will allow a user to connect to any database using any

Reviewer 3: ```suggestion
Create a role that will allow a user to connect to any database using any
```

> database account:

Reviewer: Can we use the `access` role?

Reviewer 3: @benarent Do we have any documentation for these new built-in roles yet? I couldn't find it the other day by searching "editor" or "access" using the Algolia search.

Author: No, access role doesn't have wildcard db_names and db_users.

Reviewer: Could we educate people for how to add them? ( If the below is correct? I just guessed back on `logins`) 

``` 
tctl --config=/path/to/teleport.yaml users add --roles=editor,access,db --db_name=foo --db_users=bar teleport-admin 

Access Role
```
kind: role
metadata:
  description: Access cluster resources
  id: 1617830297136031551
  name: access
spec:
  allow:
    app_labels:
      '*': '*'
    db_labels:
      '*': '*'
    db_names:
    - '{{internal.db_names}}'
    db_users:
    - '{{internal.db_users}}'
 ```

Author: @benarent Sorry not following, to add what? Access role is a part of default distribution but it just restricts db_names and db_users to `{{internal.db_names}}` as you pointed out above. If you mean to add these traits to the user, I'm not sure there's a command to do this?

Reviewer 2: @r0mant I agree with Ben on access roles, they are documented here:

https://goteleport.com/docs/access-controls/getting-started/#step-13-add-local-users-with-preset-roles

Author: @klizhentas @benarent Yes, access role would be nice here but only allows access to `{{internal.db_names/users}}` which are empty for a local user so effectively "access" role doesn't give any access to any databases.

We need some way to update traits for a local use to set those db_names/db_users. Do you folks know if there's a way to update local user traits now? I couldn't find a good way to do it.

Should we bring back flags `--db-users` and `--db-names` for "tctl users add/update" commands which would set those respective traits?

Author: Nevermind, I found a pretty good way to do it. Updated.

```sh
$ tctl --config=/path/to/teleport.yaml create <<EOF
kind: role
version: v3
metadata:
  name: db
spec:
  allow:
    db_labels:
      '*': '*'
    db_names:
    - '*'
    db_users:
    - '*'
EOF
```

> Create the user assigned the `db` role we've just created:

Reviewer 3: ```suggestion
Create a user and assign them the `db` role we've just created:
```

```sh

> $ tctl --config=/path/to/teleport.yaml users add --roles=admin,db testuser

Reviewer: ```suggestion
$ tctl --config=/path/to/teleport.yaml users add --roles=editor,acces,db teleport-admin
```

```

> ### Start Database Service with Config File

Reviewer 2: @r0mant would be cool if you had a `teleport configure --preset=db-iam` one liner instead of this config and then referenced the proper config file at the end

Author: This is a good idea. I'll do it for 6.2 if I have time and update db quickstarts, created a ticket for now: https://github.com/gravitational/teleport/issues/6368.

[RBAC](../rbac.mdx) section for more details.

To connect to a particular database server, first retrieve credentials from
Teleport using `tsh db login` command:

```sh
$ tsh db login cloudsql
```

<Admonition
  type="tip"
  title="Tip"
>
  You can be logged into multiple databases simultaneously.
</Admonition>

You can optionally specify the database name and the user to use by default
when connecting to the database instance:

```sh
$ tsh db login --db-user=teleport@<project-id>.iam --db-name=postgres cloudsql
```

<Admonition
  type="note"
  title="What database user name to use?"
>
  When connecting to the database, use the name of the database service account
  that you added as IAM database user [above](#create-service-account-for-database),
  minus the `.gserviceaccount.com` suffix. The database user name is shown on
  the Users page of your Cloud SQL instance.
</Admonition>

When logging into a PostgreSQL database, `tsh` automatically configures a section
in the [connection service file](https://www.postgresql.org/docs/current/libpq-pgservice.html)
with the name of `<cluster-name>-<database-service-name>`.

> Suppose the cluster name is "root", then you can connect to the database using

Reviewer: When using `teleport configure` we now set cluster name as the FQDN of the cluster. To make the clsuter naming less confusing should we change this to Suppose the cluster name is "teleport.example.com" .... use `psql "service=teleport.example.com-cloudsql"` ?

Author: Yeah we can. Updated.
